clang/lib/CodeGen/CGHLSLUtils.h

```diff
@@ -0,0 +1,47 @@
+
+//===----- CGHLSLUtils.h - Utility functions for HLSL CodeGen ---*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides utility functions for HLSL code generation.
+// It is used to abstract away implementation details of backends.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_LIB_CODEGEN_CGHLSLUTILS_H
+#define LLVM_CLANG_LIB_CODEGEN_CGHLSLUTILS_H
+
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/IntrinsicsDirectX.h"
+#include "llvm/IR/IntrinsicsSPIRV.h"
+
+// Define the function generator macro
+#define GENERATE_HLSL_INTRINSIC_FUNCTION(name)                                 \
+  static llvm::Intrinsic::ID get_hlsl_##name##_intrinsic(                      \
+      const llvm::Triple::ArchType Arch) {                                     \
+    switch (Arch) {                                                            \
+    case llvm::Triple::dxil:                                                   \
+      return llvm::Intrinsic::dx_##name;                                       \
+    case llvm::Triple::spirv:                                                  \
+      return llvm::Intrinsic::spv_##name;                                      \
+    default:                                                                   \
+      llvm_unreachable("Intrinsic " #name                                      \
+                       " not supported by target architecture");               \
+    }                                                                          \
+  }
+
+namespace clang::CodeGen {
+class HLSLUtils {
```

Comment:
Ish... it can also be an instance method and take no parameters since the `CGHLSLRuntime` object can check the triple itself.

Reply:
ok so from `CGHLSLRuntime::get_hlsl_all_intrinsic(CGM.getTarget().getTriple().getArch())` --> `CGM.getHLSLRuntime().get_hlsl_all_intrinsic();`

Reply:
Yea. I also wonder if there is a better way to construct the name of the functions so that they match the LLVM naming conventions.

Reply:
So kept the first character lower. as for the rest I can't camel case and keep the same name as the intrinsic.  As for underscores, the underscores for `thread_id` would look weird in the name if I didn't pick a snake case style.

Reply:
Will this only be used in cases like this where we have a unique builtin? We could use the builtin enum as a template parameter. Something like:

```c++
template<unsigned BI>
llvm::IntrinsicID getIntrinsic();

template
llvm::IntrinsicID getIntrinsic<Builtin::BI__builtin_hlsl_elementwise_all>() { ... }
```

(I probably like templates too much)

Reply:
That won't work because builtins don't have a 1 to 1 mapping with dx\spv intrinsics.

for example mad builtin. has `int_dx_imad` and `int_dx_imad` (float case is an llvm  generic intrinsics so isn't relevant).  Dot product is another example but it isn't as good of one because spirv wouldn't have the dot2 through 4 intrinsics.

dot product cases:
`int_dx_dot2`, `int_dx_dot3`, `int_dx_dot4`, `int_dx_sdot` ,`int_dx_udot`

Reply:
I'm not sure how the template makes the situation of not having 1:1 mappings between dx and spy any worse or different than your current suggestion.

The template is keyed off the builtin id not the intrinsic, and your system requires being robust to the different mappings anyways.

An alternate approach would be to add a CamelCase string to your macro definition so that we can generate CamelCase accessor methods. That would allow us to follow coding conventions with minimal impact.

Reply:
Ok one more alternative approach. If your ok with two strings  then why not  a builtin  string and the intrinsic postfix string. That would solve my one builtin to multiple intrinsics problem and the rigid naming issue.

Reply:
I don't see a problem with multiple arguments going into the macro. Using a macro like this is to reduce duplicated code, but if it is insufficiently flexible to match coding conventions and our other design considerations it isn't the right approach.

Reply:
Also `thread_id` doesn't have a builtin. So i think two names approach  might be better.

Reply:
Given that the template doesn't cover all cases, it seems like it probably adds complexity that isn't needed since all cases can be handled by the other macro approach.

Reply:
I'll remove the template